### PR TITLE
More Configurable fields and mores

### PR DESCRIPTION
**Pull request by wenfengwang**
1. More Configurable fields
2. error code of call C method returned has been processed
3. Command line tools.
4. use [logrus](https://github.com/sirupsen/logrus) as log framework.

### CODE REVIEW

File: core/api.go [hunk 1]
```diff
@@ -23,20 +23,31 @@ func Version() (version string) {
 }
 
 // NewProduer create a new producer with config
-func NewProduer(config *ProducerConfig) Producer {
+func NewProduer(config *ProducerConfig) (Producer, error) {
```
Comment (ShannonDing) [inline core/api.go]: producer, see PR 9

File: core/api.go [hunk 2]
```diff
@@ -23,20 +23,31 @@ func Version() (version string) {
...
-	GroupID     string
-	NameServer  string
-	Credentials *SessionCredentials
+	GroupID          string
+	NameServer       string
+	NameServerDomain string
+	GroupName        string
+	InstanceName     string
+	Credentials      *SessionCredentials
+
```
Comment (Hellojungle) [inline core/api.go]: I think a base interface can be extracted, such as clientConfig

File: core/api.go [hunk 3]
```diff
@@ -52,23 +63,29 @@ type Producer interface {
 }
 
 // NewPushConsumer create a new consumer with config.
-func NewPushConsumer(config *ConsumerConfig) (PushConsumer, error) {
+func NewPushConsumer(config *PushConsumerConfig) (PushConsumer, error) {
 	return newPushConsumer(config)
```
Comment (Hellojungle) [inline core/api.go]: What is the meaning? Just for make it exportable？

File: core/producer.go [hunk 4]
```diff
@@ -59,21 +60,85 @@ func (status SendStatus) String() string {
 	}
 }
 
-func newDefaultProducer(config *ProducerConfig) *defaultProducer {
+func newDefaultProducer(config *ProducerConfig) (*defaultProducer, error) {
+	if config.GroupID == "" {
```
Comment (Hellojungle) [inline core/producer.go]: Maybe judge the config with nil?

File: core/push_consumer.go [hunk 5]
```diff
@@ -89,13 +165,24 @@ func newPushConsumer(config *ConsumerConfig) (PushConsumer, error) {
 }
 
 func (c *defaultPushConsumer) Start() error {
-	C.StartPushConsumer(c.cconsumer)
+	code := C.StartPushConsumer(c.cconsumer)
```
Comment (Hellojungle) [inline core/push_consumer.go]: Maybe functions like this can be move to parent struct？

Comment (wenfengwang) [reply core/push_consumer.go]: No needed to abstract parent struct, because it simple enough.

File: core/producer.go (hunk 4 above)
Comment (wenfengwang) [reply core/producer.go]: In golang, the default value of `string` type is `""` not `nil`

File: core/api.go (hunk 1 above)
Comment (wenfengwang) [reply core/api.go]: fixed

File: core/api.go (hunk 2 above)
Comment (wenfengwang) [reply core/api.go]: fixed

File: core/producer.go (hunk 4 above)
Comment (wenfengwang) [reply core/producer.go]: fixed

File: examples/producer.go [hunk 6]
```diff
@@ -24,20 +24,23 @@ import (
...
-func SendMessage() {
-	producer := rocketmq.NewProducer(&rocketmq.ProducerConfig{GroupID: "testGroup", NameServer: "localhost:9876"})
 	producer.Start()
 	defer producer.Shutdown()
 
 	fmt.Printf("Producer: %s started... \n", producer)
 	for i := 0; i < 100; i++ {
 		msg := fmt.Sprintf("Hello RocketMQ-%d", i)
-		result := producer.SendMessageSync(&rocketmq.Message{Topic: "test", Body: msg})
+		result := producer.SendMessageSync(&rocketmq.Message{Topic: "wwf1", Body: msg})
```
Comment (ShannonDing) [inline examples/producer.go]: It is better to change the topic name back to "Test".

Comment (wenfengwang) [reply examples/producer.go]: sorry, i forgot change it back.

**Pull request merged**